Prometheus metrics for Monitoring

surenderssm opened the pull request:
**Reason for Change**:

This exposes Prometheus metrics to enable better Monitoring in MIC and NMI.

**Issue Fixed**:
<!-- If this PR fixes GitHub issue 1234, add "Fixes #1234" to the next line. -->
Fixes #263 

**Notes for Reviewers**:

Please consider this as first draft and work in progress.

kkmsft commented:
/azp run

dmeytin commented on docs/readmes/README.monitoring.md:
```diff
@@ -0,0 +1,38 @@
+
+**5. managedidentitycontroller_cycle_count**
+
+Counter that tracks the number of cycles executed in Managed Identity Controller.
```
```suggestion 
 It would be nice to have a counter how much new identities were created/deleted 
```

surenderssm replied:
@dmeytin -Thank you for this comment.We have included the suggested counters. Please have a  look.

aramase commented:
/azp run

aramase commented:
/azp run

surenderssm commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,275 @@
+		&view.View{
+			Description: AssignedIdentityAdditionDurationM.Description(),
+			Measure:     AssignedIdentityAdditionDurationM,
+			Aggregation: view.Distribution(0, 1, 2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 97),
```
Self review : distribution buckets are too large, as the duration is in seconds. Reduce the same to get more coverage.

surenderssm replied:
done.

aramase commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,275 @@
+// RecordK8SAPIOperationError records the error in KubernetesAPIOperationsErrorsCountM
+func RecordK8SAPIOperationError(operation string) {
+
+	reporter, reporterError := NewReporter()
```
Can we have this method on `reporter` receiver just like `ReportOperationAndStatus`?

surenderssm replied:
moved

aramase commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,275 @@
+}
+
+// RecordK8SAPIOperationError records the error in KubernetesAPIOperationsErrorsCountM
+func RecordK8SAPIOperationError(operation string) {
```
```suggestion
func (r *reporter) RecordK8SAPIOperationError(operation string) {
```

surenderssm replied:
This has been moved to respective package.Now reporter do not own any operation specific 
functions.

aramase commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,275 @@
+}
+
+// RecordCloudProviderOperationError records the error in CloudProviderOperationsErrorsCountM
+func RecordCloudProviderOperationError(operation string) {
```
same as above ^^

surenderssm replied:
moved.

aramase commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,275 @@
+	err := registerViews()
+
+	if err != nil {
+		log.Errorf("Failed to register views for metrics. error:%v", err)
```
If there is an error registering view for metrics, we need to return the error.

surenderssm replied:
done.

aramase commented on pkg/metrics/prometheus_exporter.go:
```diff
@@ -0,0 +1,33 @@
+)
+
+// newPrometheusExporter creates prometheus exporter and run the same on given port
+func newPrometheusExporter(namespace string, portNumber string, log log.Logger) error {
```
Don't we need to return `view.Exporter` here and register that exporter using `view.RegisterExporter`?

surenderssm replied:
Done. It was working earlier because by default Collected data will be reported via all the registered exporters.

aramase commented on pkg/crd/crd.go:
```diff
@@ -306,6 +306,15 @@ func (c *Client) RemoveAssignedIdentity(assignedIdentity *aadpodid.AzureAssigned
 	glog.V(5).Infof("Deletion %s took: %v", assignedIdentity.Name, time.Since(begin))
 	stats.Update(stats.AssignedIDDel, time.Since(begin))
+
+	metrics.NewReporterAndReport(
```
Can we add the reporter to client struct and use that for all the reports? That way we can initialize it when we create the client the first time.

surenderssm replied:
moved.

aramase commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,275 @@
+}
+
+// NewReporterAndReport creates a new instance of reporter and report given measurements
+func NewReporterAndReport(ms ...stats.Measurement) {
```
If we add the reporter into the structs where we report, I don't think we will need this function.

surenderssm replied:
Agree,removed.

surenderssm commented on pkg/nmi/server/server.go:
```diff
@@ -178,6 +180,17 @@ func (fn appHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	latency := time.Since(start)
 	logger.Infof("Status (%d) took %d ns", rw.statusCode, latency.Nanoseconds())
+
+	reporter, reporterError := metrics.NewReporter()
```
moved to server struct.

surenderssm commented on pkg/mic/mic.go:
```diff
@@ -248,6 +251,13 @@ func (c *Client) Sync(exit <-chan struct{}) {
 	totalWorkDoneCycles := 0
 	totalSyncCycles := 0
+
+	reporter, reporterError := metrics.NewReporter()
```
moved to client struct and error is being returned if we fail to create the reporter.

surenderssm commented on pkg/crd/crd.go:
```diff
@@ -522,5 +537,13 @@ func (c *Client) UpdateAzureAssignedIdentityStatus(assignedIdentity *aadpodid.Az
 	glog.V(5).Infof("Patch of %s took: %v", assignedIdentity.Name, time.Since(begin))
 
+	if err != nil {
+		recordError("update_azure_assigned_identity_Status")
```
done.

surenderssm commented:
@aramase - Thank you for the comments.

- cloud_provider_operations_duration_seconds metric has been added to track time taken for cloud operations.

- metrics package has been refactored with suggestions

Have a look let me know you thoughts on changes.

kkmsft commented on pkg/cloudprovider/vm.go:
```diff
@@ -36,8 +44,16 @@ func NewVirtualMachinesClient(config config.AzureConfig, spt *adal.ServicePrinci
 	client.AddToUserAgent(version.GetUserAgent("MIC", version.MICVersion))
 
+	reporter, err := metrics.NewReporter()
+
```
nit: extra space

surenderssm replied:
Noted

surenderssm replied:
@kkmsft fixed.

kkmsft commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,254 @@
+
+// ReportOperation records given measurement by operation type.
+func (r *Reporter) ReportOperation(operationType string, measurement stats.Measurement) error {
+
```
nit: extraneous space.

kkmsft commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,254 @@
+
+// RegisterAndExport register the views for the measures and expose via prometheus exporter
+func RegisterAndExport(port string, log log.Logger) error {
+
```
nit: extraneous space.

surenderssm replied:
done

kkmsft commented:
@surenderssm - can you please add set of UTs and e2e to test the metrics ?

aramase commented:
/azp run

DanielShor commented on pkg/cloudprovider/cloudprovider.go:
```diff
@@ -337,3 +338,22 @@ func ParseResourceID(resourceID string) (azure.Resource, error) {
 }
+
+// recordError records the error in appropriate metric
+func recordError(reporter *metrics.Reporter, operation string) {
```
There is a weird behavior with this metric. On the one hand, when looking on `aadpodidentity_nodemanagedidentity_operations_duration_seconds_bucket` metric, we can see a lot of requests (with operation_type=`/metadata/identity/oauth2/token`) with status_code= 403. On the other hand, we have no data at all from `aadpodidentity_cloud_provider_operations_errors_count` metric. Would appreciate your help,

elutsky replied:
I understand that part of this 403 is wired noise, but I think it is very important to be able to get "quality" 403, without a noise - should allow to monitor and rise alert.

surenderssm replied:
@DanielShor , @elutsky - 
As per my understanding following are the scenarios where NMI will return 403.

1. Failed to acquire a token using the MSI
2. No matched identity in CREATED state was found
3. Request blocked by AAD Pod Identity NMI

"No data at all from aadpodidentity_cloud_provider_operations_errors_count metric"
Only (1) from above 'Failed to acquire token via MSI', falls in the bucket of cloud_provider_operations and is not being tracked.
I will take up task of adding MSI token operations to cloud_provider_operations metric.

Thank you for finding this gap and bringing to our notice.
@kkmsft / @aramase - Any other thoughts on this ?

surenderssm replied:
@DanielShor , @elutsky , @kkmsft / @aramase - Added MSI token operations to cloud_provider_operations metric

DanielShor commented on pkg/metrics/metrics.go:
```diff
@@ -0,0 +1,254 @@
+		nodeManagedIdentityOperationsDurationName,
+		"Duration in seconds of node managed identity operations",
+		stats.UnitMilliseconds)
+	// "operation_type", "status_code"
```
Can we add some labels to this metric? I would suggest the following:
- AzureIdentity name (k8s object) which related to this request
- K8s namespace of the AzureIdentity 
- Azure resource ID (of which we are trying to get token for)

We are dealing with several namespaces in every cluster, and expect a lot of different identities defined. These labels can help us a lot with managing all of these objects. 
WDYT?

surenderssm replied:
@danielShor - Agree with your inputs.

Discussed metric NodeManagedIdentityOperationsDurationM, is being recorded in function ServeHTTP (for both success as well as error paths) in server.go(nmi). In this context we don't have details of identity but we do have reference of *http.Request and following details can be retrieved.
- NameSpace : Can be retrieved from header : "podns"
- AzureResource: Can be retrieved from query parameters : key :"resource"

Let me know if we can proceed only with these two new labels ?

DanielShor replied:
@surenderssm, these 2 are great. Can we also get the AzureIdentity name (k8s object) which related to this request? (I know it might be complicated, but can be very helpful). If too complicated for now, and may take some time we will be happy with these 2 labels :)

surenderssm replied:
@danielshor - Added namespace,resource label to NodeManagedIdentityOperationsDurationM metric. We will analyse and decide on AzureIdentity label in future iterations.

DanielShor commented:
Hi @surenderssm , I am testing AAD Pod Identity project in our workloads for about 2 months. This is PR is really important for us, so it is very much appreciated. 
I have been trying an image which contains this PR features, and overall this is great! 
I added 2 comments, can you please take a look? 
Moreover, Is it possible to add metric which track the number of identities exist in the cluster, broken down by K8s namespace? (can be Gauge for example)

Would love to hear your thoughts. Please lMK if we can help. 
Thanks again for the PR.

surenderssm commented:
@DanielShor , @elutsky - 

Thank you for your review and trying out the changes at your side.
I have replied on your comments, please let me know your thoughts on the same.

> Moreover, Is it possible to add metric which track the number of identities exist in the cluster, broken down by K8s namespace? (can be Gauge for example)

Available metrics "assigned_identity_addition_count", "assigned_identity_deletion_count" can be inferred to get insights around number of identities.
 
I do see value in having a separate metric, Let me add metric with number of identities broken down by namespace. This will be emitted every time a cycle completes in mic.

DanielShor commented:
@surenderssm, Thanks for responding! This PR is really important for us, not sure we can wait more with the needed changes. When UnitTests are in place, we are OK with what's currently coded, and will talk about the changes later on. Sounds OK?

surenderssm commented:
@DanielShor , @aramase , @kkmsft -

1. Added adal auth operations to CloudProviderOperations metrics
2. Added ns,resource to NodeManagedIdentityOperationsDurationM metric
3. Added unit tests for metrics

Please have a look and let me know your inputs on the changes.

aramase commented:
/azp run

aramase commented on pkg/auth/auth.go:
```diff
@@ -2,66 +2,83 @@ package auth
 	}
 	token := spt.Token()
-
+	recordDuration(adalTokenOperationName, time.Since(begin))
```
should this be `adalTokenFromMSIOperationName`?

surenderssm replied:
done.

aramase commented:
/azp run

kkmsft commented on pkg/auth/auth.go:
```diff
@@ -90,3 +109,26 @@ func init() {
+}
+
+// recordError records the error in appropriate metric
+func recordError(operation string) {
```
Can we simplify this by having one single call in the metrics package and all other modules just call in with the counters required in a follow up PR.

surenderssm replied:
metrics package has 'ReportOperation','Report' to record given measurements.
Internally same is used by recordError.It has been abstracted by recordError to avoid code cluttering and to refer particular measurement at single place for any future changes.

Following are the options we can think of .
1. Keep it as-is
2. Move ReportCloudProviderOperations() to metrics package
3. Use reporter.ReportOperation with measurement name instead of recordError

@kkmsft  - Let me know which options sounds good ?

`func recordError(operation string) {
	if reporter != nil {
		reporter.ReportOperation(
			operation,
			metrics.CloudProviderOperationsErrorsCountM.M(1))
	}
}`

kkmsft commented on pkg/cloudprovider/cloudprovider.go:
```diff
@@ -337,3 +338,22 @@ func ParseResourceID(resourceID string) (azure.Resource, error) {
 }
+
+// recordError records the error in appropriate metric
+func recordError(reporter *metrics.Reporter, operation string) {
```
Same here. please.

surenderssm replied:
Whatever approach we freeze in above, will be followed here.

kkmsft commented on pkg/cloudprovider/vm.go:
```diff
@@ -14,8 +15,15 @@ import (
 type VMClient struct {
-	client compute.VirtualMachinesClient
+	client   compute.VirtualMachinesClient
+	reporter *metrics.Reporter
```
Cant we have a static declaration of metrics and call it directly from all the places without having to initialize for every package ?

surenderssm replied:
yes, we can think of something like  `metrics.Reporter` which can be initialized via `init()` in metrics package. All other packages can use directly metrics.Reporter to report any metric.
Will remove reference of `reporter *metrics.Reporter` from the structs like `VMClient /server` and others

kkmsft commented on pkg/cloudprovider/vm.go:
```diff
@@ -14,8 +15,15 @@ import (
 )
 
+const (
+	getVmOperationName = "vm_get"
```
Can we please move this to single package location ?

surenderssm replied:
Shouldn't we keep operations name local to package wherever they are used ?
Shall we move all operations name to package metrics ?

surenderssm commented on pkg/auth/auth.go:
```diff
@@ -2,66 +2,83 @@ package auth
 	// Get the MSI endpoint accoriding with the OS (Linux/Windows)
 	msiEndpoint, err := adal.GetMSIVMEndpoint()
 	if err != nil {
+		recordError(adalTokenFromMSIOperationName)
```
agree, will take care in next iteration

surenderssm commented on pkg/auth/auth.go:
```diff
@@ -2,66 +2,83 @@ package auth
 	}
 	token := spt.Token()
-
+	recordDuration(adalTokenFromMSIOperationName, time.Since(begin))
```
Agree.

aelmanaa commented:
hello @surenderssm are prometheus metrics documented somewhere?

Freakazoid182 commented:
> 
> 
> hello @surenderssm are prometheus metrics documented somewhere?

https://github.com/Azure/aad-pod-identity/blob/master/website/content/en/docs/Configure/prometheus_monitoring.md

aramase commented:
The official documentation has been moved to a docs site - https://azure.github.io/aad-pod-identity/docs/configure/prometheus_monitoring/
